src/analytics/mod.rs

```diff
@@ -3,5 +3,181 @@
 
 //! Various analytics that give insight into the usage of the tangle.
 
+use futures::TryStreamExt;
+
+use self::{
+    ledger::{
+        AddressActivity, AddressBalanceAnalytics, BaseTokenActivityAnalytics, LedgerOutputAnalytics,
+        TransactionAnalytics, UnclaimedTokenAnalytics,
+    },
+    tangle::{BlockActivityAnalytics, BlockAnalytics},
+};
+use crate::{
+    db::{
+        collections::analytics::{Measurement, PerMilestone, TimeInterval},
+        influxdb::InfluxDb,
+    },
+    tangle::{BlockData, InputSource, LedgerUpdateStore, Milestone},
+    types::{
+        ledger::BlockMetadata,
+        stardust::block::{payload::TransactionEssence, Block, Input, Payload},
+        tangle::ProtocolParameters,
+    },
+};
+
 pub mod ledger;
 pub mod tangle;
+
+#[allow(missing_docs)]
+pub enum Analytic {
+    AddressBalance(AddressBalanceAnalytics),
+    BaseTokenActivity(BaseTokenActivityAnalytics),
+    BlockActivity(BlockActivityAnalytics),
+    DailyActiveAddresses(AddressActivity),
+    LedgerOutputs(LedgerOutputAnalytics),
+    LedgerSize,
+    OutputActivity,
+    ProtocolParameters(ProtocolParameters),
+    UnclaimedTokens(UnclaimedTokenAnalytics),
+    UnlockConditions,
+}
+
+impl<'a, I: InputSource> Milestone<'a, I> {
+    /// Update a list of analytics with this milestone
+    pub async fn update_analytics(&self, analytics: &mut [Analytic], influxdb: &InfluxDb) -> eyre::Result<()> {
+        let mut cone_stream = self.cone_stream().await?;
+        let ledger_updates = self.ledger_updates().await?;
+
+        self.begin_milestone(analytics);
+
+        while let Some(BlockData { block, metadata, .. }) = cone_stream.try_next().await? {
+            self.handle_block(analytics, &block, &metadata, &ledger_updates)
+        }
+
+        self.end_milestone(analytics, influxdb).await?;
+
+        Ok(())
+    }
+
+    fn begin_milestone(&self, analytics: &mut [Analytic]) {
+        for analytic in analytics {
+            match analytic {
+                Analytic::AddressBalance(stat) => stat.begin_milestone(self.at),
+                Analytic::BaseTokenActivity(stat) => stat.begin_milestone(self.at),
+                Analytic::BlockActivity(stat) => stat.begin_milestone(self.at.milestone_index),
+                Analytic::DailyActiveAddresses(stat) => stat.begin_milestone(self.at),
+                Analytic::LedgerOutputs(stat) => stat.begin_milestone(self.at),
+                Analytic::LedgerSize => todo!(),
+                Analytic::OutputActivity => todo!(),
+                Analytic::ProtocolParameters(params) => {
+                    if params != &self.protocol_params {
+                        *params = self.protocol_params.clone();
+                        // TODO: either signal that we should write this at the end or do it now
+                        // TODO: re-init all of the analytics, what is this going to take? Can we even do it here?
```

Comment:
```suggestion
                        // TODO: re-init all of the analytics, what is this going to take? Can we even do it here?
                        unimplemented!()
```

Reply:
Oh, right, I forgot about this comment. It's already implemented, just not here.